collector/topological_inventory.py

```diff
@@ -173,30 +175,111 @@ def _query_sub_collection(
     return all_data
 
 
-def worker(_: str, source_id: str, dest: str, b64_identity: str) -> None:
-    """Worker for topological inventory.
+def worker(
+        _: str,
+        tenants_url: str,
+        all_tenants: bool,
+        source_id: str,
+        dest: str,
+        b64_identity: str
+) -> None:
+    """Topological inventory Worker.
 
     Parameters
     ----------
     _ (str)
         Skipped
+    tenants_url (str)
+        URL for retrieving Tenants
+    all_tenants (bool)
+        Flag to indicate whether to retrieve data for ALL Tenants
     source_id (str)
         Job identifier
     dest (str)
         URL where to pass data
     b64_identity (str)
         Red Hat Identity base64 string
 
+
     """
     thread = current_thread()
     LOGGER.debug('%s: Worker started', thread.name)
 
+    headers = {"x-rh-identity": b64_identity}
+
+    if all_tenants:
+        resp = utils.retryable('get', tenants_url, headers=headers)
+        tenants = resp.json()
+        LOGGER.info(
+            'Fetching data for ALL(%s) Tenants',
+            len(tenants)
+        )
+        for tenant in tenants:
+            topological_inventory_data(
+                _,
+                source_id,
+                dest,
+                tenant["external_tenant"],
+                thread
+            )
+    else:
+        tenant = json.loads(base64.b64decode(b64_identity))["identity"]
+        LOGGER.info(
+            'Fetching data for current Tenant:[%s]',
+            tenant["account_number"]
+        )
+        topological_inventory_data(
+            _,
+            source_id,
+            dest,
+            tenant["account_number"],
+            thread
+        )
+
+    LOGGER.debug('%s: Done, exiting', thread.name)
+
+
+def topological_inventory_data(
+        _: str,
+        source_id: str,
+        dest: str,
+        tenant: str,
+        thread
+) -> None:
+    """Generate Tenant data for topological inventory.
+
+    Parameters
+    ----------
+    _ (str)
+        Skipped
+    source_id (str)
+        Job identifier
+    dest (str)
+        URL where to pass data
+    tenant (str)
+        Tenant Account Number
+    thread
+        Current Thread
+
+    """
+    header = {
+        "identity":
+            {
+                "account_number": tenant
+            }
+    }
+
+    b64_tenant_identity = base64.b64encode(
```

Comment:
Here, you're encoding something you've decoded one step above. Why don't you pass the data directly then? You already have the headers when this func is called with one tenant. If it's called with multiple tenants, you can encode the response beforehead.

Reply:
You're duplicating code. It's really tangled. Why don't you simply do the same as you did before in the consumer? It's the same problem in a nutshell. https://github.com/ManageIQ/aiops-consumer/pull/6

Whole this PR can be as simple as:
```python
def tenant_to_header(account_number):
     rh_identity = dict(identity=dict(account_number=account_number)
     b64_identity = [base64.b64encode(json.dumps(rh_identity).encode())
     return {'x-rh-identity': b64_identity)
...


headers = [{"x-rh-identity": b64_identity}]

if all_tenants:
    resp = utils.retryable('get', tenants_url, headers=headers)
    headers = [tenant_to_headers(t["external_tenant"]) for t in resp.json()]

for h in headers:
     topological_inventory_data(_, source_id, dest, h, thread)
```

Reply:
> Here, you're encoding something you've decoded one step above

The main reason for decoding was to retrieve the account number and display it in the log.

Logging, I would say, is probably a key feature for this PR.
With multiple tenants/accounts and so many queries, retrieving the data can be a very daunting operation. Hence I think it is important to log the account numbers systematically.

Although, I think it is OK if we don't log the account number for the single-tenant use case.
So I will be removing this piece of code
`tenant = json.loads(base64.b64decode(b64_identity))["identity"]`
which is the decoding part

(single-tenant use case might be more handy in a DEV setup anyway and may not be used in the prod setting)